# Detection of long non-coding RNA homology
datasets and scripts adopted in the paper "Detection of long non-coding RNA homology, a comparative study on alignment and alignment-free metrics"


### Please install the following Python package and C++ source before proceeding:
 - decaf+py from http://bioinformatics.org.au/tools/decaf+py/
 - kmacs from http://kmacs.gobics.de/

 References
 - HÃ¶hl, M., Rigoutsos, I., & Ragan, M. A. (2006). Pattern-based phylogenetic distance estimation and tree reconstruction. Evolutionary Bioinformatics, 2, 117693430600200016.
 - C.-A. Leimeister, B. Morgenstern (2014). kmacs: the k-Mismatch Average Common Substring Approach to alignment-free sequence comparison. Bioinformatics 30, 2000-2008
 - S. Horwege, S. Lindner, M. Boden, K. Hatje, M. Kollmar, C.-A. Leimeister, B. Morgenstern (2014). Spaced words and kmacs: fast alignment-free sequence comparison based on inexact word matches. Nuc. Acids Research 42, W7-W11

